bolt-design-system/bolt
# Major Jest Upgrades

![8cd45a76-aa0a-44d1-94e9-ff84b04c0f62](https://user-images.githubusercontent.com/1617209/53090142-80364300-34dc-11e9-8221-a9669b8e39bb.jpeg)

## Jira
http://vjira2:8080/browse/BDS-1041

## Summary
Major upgrades to Jest to support testing web component rendering in a browser and web page instance that's generated automatically, in addition to adding the ability to run visual regression tests on these.

## Details

### Jest-specific Updates
- Updates global Jest config to automatically skip over `*.e2e.js` and `*.data.js` test files
- Added automatic transpiling of ES6 code (via Babel) to allow component tests to use modern JS libraries
- Polyfilled / updated the testing environment in order to use helper libraries for testing web components
- Wired up Jest to automatically boot up webpack dev server to remove the need to compile Pattern Lab before running any Jest tests
  -  Added new webpack dev server setup for quickly testing components using JIT (just in time) compiling
- Added basic visual regression testing to Jest via [Jest Image Snapshot](https://www.npmjs.com/package/jest-image-snapshot) + added [jest-dom](https://www.npmjs.com/package/jest-dom) as a new tool to help evaluate the state of the DOM inside components being tested
  - Also added simple image screenshot test reporting functionality to Jest to assist with debugging VRT issues until full Github integration is wired up 
  - Failed tests automatically get their screenshots uploaded to now.sh from Travis to allow for simple troubleshooting 
- **Faster Twig Renderer**:
  - Update Twig renderer to include external method to stop currently running server
  - Updated Twig renderer calls in button component example to compile even faster

### Jest Test-specific Updates
- Renamed the figure component test's external data file to automatically get excluded
- Removed the old button component JS test
- Wired up the existing button component's Jest tests to now include two visual regression tests in addition to to web component test examples
- Moved the image component's test file to the proper folder
- Updated the monorepo Jest test to speed things up by 20 seconds!

## How to test
- Review updates to understand the changes this includes -- especially the main jest config file + the button component's test updates!
- Confirm Jest tests running as expected
- Confirm the full build runs and deploys as expected
